### PR TITLE
[stdlib] Resolve unsafeBitCast warnings in Runtime.swift.gyb.

### PR DESCRIPTION
**Pull request by jrose-apple**
No intended functionality change.

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please benchmark

**Review by moiseev — APPROVED** [review]
LGTM. Thanks, @jrose-apple !

File: stdlib/public/core/Runtime.swift.gyb [hunk 1]
```diff
@@ -72,10 +70,20 @@ func _stdlib_atomicCompareExchangeStrongPtr<T>(
...
+    capacity: 1
+  ) { target in
+    return expected.withMemoryRebound(
+      to: Optional<UnsafeRawPointer>.self,
+      capacity: 1
+    ) { expected in
+      return _stdlib_atomicCompareExchangeStrongPtr(
+        object: target,
+        expected: expected,
+        desired: UnsafeRawPointer(desired))
```
Comment (atrick) [inline stdlib/public/core/Runtime.swift.gyb]: I'm not sure why `desired` was originally bitcast, nor why it needs a seemingly redundant initializer call here.

**Review by atrick — COMMENTED** [review]
It's a good cleanup. I'm concerned about adding withMemoryRebound closures to atomic operations. It's not the sort of place we want heap allocation. In the atomicFetch cast, just use `assumingMemoryBound(to:)`. `Int` is an `Int64`. I guess we can ignore the performance of the generic entry point unless the benchmark results look bad.

**Comment by atrick** [comment]
Mandatory inlining should actually get rid of the closure. But still, assumingMemoryBound(to:) is perfectly fine for UnsafePointer<Int> -> UnsafePointer<Int64>. And in general it's fine if the pointer is only used inside a C function.

File: stdlib/public/core/Runtime.swift.gyb (hunk 1 above)
Comment (jrose-apple) [inline stdlib/public/core/Runtime.swift.gyb]: Purely style. I don't like relying on pointer-to-pointer conversions in pure Swift code.

**Comment by jrose-apple** [comment]
There's no assumingMemoryBound on the typed pointers, which makes it a little more annoying, but I can switch to that if you prefer.

**Comment by atrick** [comment]
In my mind, an atomic increment is more primitive than a closure taking function. I don't like unnecessary abstraction in the implementation of language primitives. I've seen too much SIL code and waiting too long for builds.

I confess, I was thinking that `Int, Int64` were typealiases on a 64-bit platform, which is incorrect.

It's still safe to use `assumingMemoryBound`. In fact, that's no less safe than `unsafeBitCast` was. But proving correctness depends on the fact that `Int` and `Int64` are imported from C for the purpose of alias analysis. In other words, they need to be special-cased as "related types" anyway for C interoperability to be safe.

**Comment by atrick** [comment]
Also, none of this is meant to be pretty or convenient. I wouldn't worry about elegance or best practices at this level. Yes, we should for stdlib data structures, but this is just a builtin that happens to be implementable in Swift.

**Comment by jrose-apple** [comment]
Updated. @atrick, is this what you were thinking? Are all of these uses safe? (Compare with the previous version: jrose-apple/swift@/9fd029820d2868e140bd96d42d3d4eae86af3f33)

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by atrick** [comment]
Yes, I'll stand by the correctness of this patch...

I had to think about it overnight. Your original patch was generally the right way to handle these casts. But since this is what I call a "language primitive" it's a good place to explore the boundaries of manual optimization, both as a performance guarantee and because adding abstraction in "inlineable" primitives could conceivably hurt build time.

I *was* arguing that `Int` is related to `Int64` on 64-bit. That's not naturally true though. Even in C, `long` and `long long` are not AFAIK "compatible" types (that's analogous with Swift's definition of related types). @rjmccall  could weigh in here. However, I recommend that we explicitly call-out `Int` as related to `Int32`/`Int64` depending on the platform. That way `#if arch` code like this will be sane and consistent with the intuition of users. I would never want the compiler to disambiguate these cases.

The rule in the compiler should actually fall out of this logic:

`struct Int` and `struct Int64` represent C types in imports, so alias
analysis needs to look at their members to determine whether pointers
to them can alias. Both just hold a `Builtin.Int64`, so they are
related by virtue of being structurally compatible C types, and their
pointers might alias.

I was not originally suggesting you change the other cases, but as long as we're being maximally aggressive in the atomics implementation... I also claim that we should explicitly call out `AnyObject`, `UnsafeRawPointer`, `UnsafePointer<T>`, and their optional variants as related types. I have seen other code that makes this assumption and already have a note to add that case to the specs.

`UnsafeRawPointer` and `UnsafePointer<T>` are both, for the purpose of alias analysis, imported from C, and both hold `Builtin.RawPointer`. So they are related by the same logic a `Int` and `Int64`.

I suggest we call out `AnyObject` in the type system spec to be a magic protocol that is by fiat related to `UnsafeRawPointer`, `UnsafePointer<T>` and any other class or class protocol. That way, pointers-to-generic-pointers may alias.

`UnsafePointer<T>` is still unrelated `UnsafePointer<U>` and arbitrary class protocols are unrelated to each other.

**Comment by jckarter** [comment]
Can we sidestep the memory binding questions by having the `_swift_stdlib_atomic*` wrappers take an `UnsafeRawPointer`? As you noted, @atrick, the underlying LLVM operations are untyped operations on `Builtin.RawPointer`, so requiring the memory to be bound to a type seems unnecessary.

**Comment by atrick** [comment]
It depends on how clear you want the wrapper API's to be. To use `_stdlib_atomicCompareExchange` correctly, it's important to realize that you're passing a pointer-to-pointer. To use `_stdlib_AtomicFetchAndAdd` it's important to realize you're passing a pointer-to-Int.

The `assumingMemoryBound` cast could be removed from both the generic `_stdlib_atomicCompareExchangeStrongPtr<T>` and the `ArcRef` wrapper by introducing an even lower level `Impl` helper that just takes `UnsafeRawPointer` and does an untyped pointer load of the expected value from that.

@jrose-apple can decide if that's worth it. Or just hand it to me if you're bored of this.

**Comment by jrose-apple** [comment]
I tried for a moment and it didn't look like it was going to go very well due to the need to load from and store to `expected`. I'll let you make the final call.

@swift-ci Please smoke test

**Comment by atrick** [comment]
@jrose-apple I won't get to this today so I'll go ahead and push your change to fix the warning. I made a note to come back to it later. Thanks!

**Pull request merged**